api/dp-openapi.yaml

```diff
@@ -591,6 +591,354 @@ paths:
+              schema:
+                $ref: '#/components/schemas/round_data'
+        404:
+          $ref: '#/components/responses/NotFound'
+        429:
+          $ref: '#/components/responses/RateLimitExceeded'
+        default:
+          $ref: '#/components/responses/UnexpectedError'
+
+  /proofrounds/{proofroundid}/proofprojects/{proofprojectid}/proofpages:
+    post:
+      tags:
+      - project
+      description: Checks out a page for proofreading
+      parameters:
+      - name: proofroundid
+        in: path
+        description: The round identifier (as defined by array_keys($Round_for_round_id_)
+        required: true
+        schema:
+          type: string
+      - name: proofprojectid
+        in: path
+        description: The project id
+        required: true
+        schema:
+          type: string
+      responses:
+        200:
+          description: Page data
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/proof_page'
+        401:
+          $ref: '#/components/responses/Unauthorized'
+        404:
+          $ref: '#/components/responses/NotFound'
+        429:
+          $ref: '#/components/responses/RateLimitExceeded'
+        default:
+          $ref: '#/components/responses/UnexpectedError'
+
+  /proofrounds/{proofroundid}/proofprojects/{proofprojectid}/proofpages/{proofpage}:
+    put:
+      tags:
+      - project
+      description: save the text and/or possibly change state
+      parameters:
+      - name: proofroundid
+        in: path
+        description: The round identifier (as defined by array_keys($Round_for_round_id_)
+        required: true
+        schema:
+          type: string
+      - name: proofprojectid
+        in: path
+        description: The project id
+        required: true
+        schema:
+          type: string
+      - name: proofpage
+        in: path
+        description: the name of the page, must be in temp state and owned by user
+        required: true
+        schema:
+          type: string
+      requestBody:
+        description: data for operation
+        required: true
+        content:
+          application/json:
+            schema:
+              oneOf:
+                - $ref: '#/components/schemas/page_save'
+                - $ref: '#/components/schemas/page_return'
+                - $ref: '#/components/schemas/page_mark_bad'
+      responses:
+        200:
+          description: operation success
+          content: {}
+        401:
+          $ref: '#/components/responses/Unauthorized'
+        404:
+          $ref: '#/components/responses/NotFound'
+        422:
+          $ref: '#/components/responses/InvalidCharacters'
+        429:
+          $ref: '#/components/responses/RateLimitExceeded'
+        default:
+          $ref: '#/components/responses/UnexpectedError'
+
+  /proofrounds/{proofroundid}/proofprojects/{proofprojectid}/ownedpages/{ownedpage}:
```

Comment:
I don't understand what `ownedpages` is vs `proofpages` - can you clarify how these differ and where this one is used?

Reply:
The difference is in the page state. (Assuming we are no longer using the "out" state.) A proofpage just means that it is in the "temp" state and owned by the user.
A proofpage can then be saved as temp (no state chage) or "done" (changes to a "savedpage"), returned to the round (change to "avail", no longer a "proofpage"), or marked as bad (no longer a "proofpage"). None of these require any data to be returned to the client unless something is wrong (such as bad characters or limit exceeded for saving.
I think it would really be logical to define a "savedpage" which is in "saved" state and owned by the user. The only thing that the user can do to a page in this state is to re-open it, that is convert it to a "proofpage" which changes its state to "temp" and make some other database changes. This operation returns page data to the client.
The exising code makes links to pages which can be either saved or temp so in order to avoid adding more complexity to the interface we can define an "ownedpage" which can be in either state. If we "reopen" it, the state may or may not change. (Similar to what LPage:resume_saved_page() does).

Reply:
If `ownedpages` and `proofpages` are based on states of a page then they aren't separate resources -- the resources is the page -- and should be selectors to an API rather than separate endpoints.

Reply:
We can define a resource in a more abstract way:  a "proofpage" is defined as a page checked out to a user and in "temp" state. We can create this resource by the POST checkout operation. If we return it to the round the resource ceases to exist. If we save it as "done" the "proofpage" also ceases to exist but is converted to an "ownedpage". It is easy to make validators to check whether these resources exist.